async-nats/src/lib.rs

```diff
@@ -903,8 +907,50 @@ impl Subscriber {
     ///  subscriber.unsubscribe();
     /// # Ok(())
     /// # }
-    pub fn unsubscribe(self) {
-        drop(self)
+    pub async fn unsubscribe(&mut self) {
+        self.receiver.close();
+        self.sender
+            .send(ClientOp::Unsubscribe { id: self.uid })
+            .await
+            .ok();
+    }
+
+    /// Unsubscribes from subscription after reaching given number of messages.
+    /// This is the total number of messages received by this subcsription in it's whole
+    /// lifespan. If it already reeached or surpassed the passed value, it will immediately stop.
+    ///
+    /// # Examples
+    /// ```
+    /// # use futures_util::StreamExt;
+    /// # #[tokio::main]
+    /// # async fn unsubscribe() -> Result<(), Box<dyn std::error::Error>> {
+    /// let mut client = async_nats::connect("demo.nats.io").await?;
+    ///
+    /// for _ in 0..3 {
+    ///     client.publish("test".into(), "data".into()).await?;
+    /// }
+    ///
+    /// let mut sub = client.subscribe("test".into()).await?;
+    /// sub.unsubscribe_after(3).await;
+    /// client.flush().await?;
+    ///
+    /// while let Some(message) = sub.next().await {
+    ///     println!("message received: {:?}", message);
+    /// }
+    /// println!("no more messages, unsubscribed");
+    /// # Ok(())
+    /// # }
+    pub async fn unsubscribe_after(&mut self, unsub_after: u64) {
+        self.max = Some(unsub_after);
```

Comment:
I would forego tracking anything here, and instead just send the unsub message with the optional limit parameter.

Reply:
Haven't realized that usub has that additional param and it's not handled purely client side.

Go client does that counting and that is probably relevant when resubscribing?
Will have to dig the Go code little more.

Reply:
For client side, since it is a stream one can do `subscriber.limit(n)` to get N messages before doing something else.